lib/madwf_tensor.cu

```diff
@@ -0,0 +1,78 @@
+#include <color_spinor_field.h>
+#include <tunable_nd.h>
+#include <device_vector.h>
+#include <madwf_transfer.h>
+#include <kernels/madwf_tensor.cuh>
+#include <tunable_reduction.h>
+#include <madwf_ml.h>
+
+namespace quda
+{
+  namespace madwf_ml
+  {
+
+    template <class storage_t> class tensor_5D_wrapper : public TunableMultiReduction<1>
+    {
+      const ColorSpinorField &x;
+      const ColorSpinorField &y;
+      using Arg = Tensor5DReduceArg<storage_t>;
+      using reduce_t = typename Arg::reduce_t;
+      reduce_t *wm_p;
+
+    private:
+      unsigned int minThreads() const { return x.VolumeCB() / x.X(4); }
+
+    public:
+      tensor_5D_wrapper(const ColorSpinorField &x, const ColorSpinorField &y, MadwfAcc::transfer_float *wm_p) :
+        TunableMultiReduction(x, x.X(4) * y.X(4)), x(x), y(y), wm_p(reinterpret_cast<reduce_t *>(wm_p))
+      {
+        char tmp[512];
+        sprintf(tmp, ",%02d->%02d", y.X(4), x.X(4));
```

Comment:
Using `sprintf` for int to string conversion is very slow.  Use `u32toa` instead (uint_to_char.h).

Reply:
My preference would be to not use `sprintf` at all 😉  It's probably worth looking at `std::string` instead of `char*`, and also we could look into fmtlib (https://github.com/lattice/quda/issues/1176) for conversion, since it looks to be fast and is the basis for C++20 `std::format`.  For now, as long we don't use `sprintf` for integer to string, I'm happy.

Reply:
![n9n6sqvOnVY51yQX8ciX-BTJTsw](https://user-images.githubusercontent.com/12970356/148278757-fdb2f2db-9f17-4be0-8f23-9ffeb020f8c4.png)
I did a quick benchmark of 4 different options:
- Using `sprintf("%d")` (the original one, `Sprintf_d`)
- Using `i32toa` and `sprintf("%s")` (`Sprintf_s_i32toa`)
- Using `i32toa` and `strcat` (`Strcat_i32toa`)
- Using `i32toa` and `std::string` (`StdString_i32toa`)
Seems we should avoid `sprintf`.

Reply:
https://github.com/lattice/quda/pull/1061/commits/703f9d363082c78607d77ac0d576e66733264f35